mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,118 +19,555 @@
+            k_new_pending_merge_commit=FLUSH_PENDING_MERGE_COMMIT
+        else
+            k_mainline_commit=$(kube-commit ${f_mainline_commit})
+
+            # check under which merge with the mainline ${k_mainline_commit}) is
+            k_new_pending_merge_commit=$(git-find-merge ${k_mainline_commit} upstream-branch)
+            if [ "${k_new_pending_merge_commit}" = "${k_mainline_commit}" ]; then
+                # it's on the mainline itself, no merge above it
+                k_new_pending_merge_commit=""
+            fi
+            if [ ${dst_branch} != master ] && is-merge-with-master "${k_mainline_commit}"; then
+                # merges with master on non-master branches we always handle as pending merge commit.
+                k_new_pending_merge_commit=${k_mainline_commit}
+            fi
         fi
-        if echo ${commits_change_godeps} | grep -q ${commit} > /dev/null; then
-            echo "branch commit ${commit} changes Godeps.json"
-            cherrypicks_change_godeps="true"
+        if [ -n "${k_pending_merge_commit}" ] && [ "${k_new_pending_merge_commit}" != "${k_pending_merge_commit}" ]; then
+            # the new pending merge commit is different than the old one. Apply the old one. Three cases:
+            # a) it's a merge with master on a non-master branch
+            #    (i) it's on the filtered-branch
+            #    (ii) it's dropped on the filtered-branch, i.e. fast-forward
+            # b) it's another merge
+            local dst_parent2="HEAD"
+            if [ ${dst_branch} != master ] && is-merge-with-master "${k_pending_merge_commit}"; then
+                # it's a merge with master. Recreate this merge on ${dst_branch} with ${dst_parent2} as second parent on the master branch
+                local k_parent2="$(git rev-parse ${k_pending_merge_commit}^2)"
+                read k_parent2 dst_parent2 <<<$(look -b ${k_parent2} ../kube-commits-$(basename "${PWD}")-master)
+                if [ -z "${dst_parent2}" ]; then
+                    echo "Corresponding k8s.io/$(dirname ${PWD}) master branch commit not found for upstream master merge ${k_pending_merge_commit}. Odd."
+                    return 1
+                fi
+
+                f_pending_merge_commit=$(branch-commit ${k_pending_merge_commit} filtered-branch)
+                if [ -n "${f_pending_merge_commit}" ]; then
+                    echo "Cherry-picking k8s.io/kubernetes master-merge  ${k_pending_merge_commit}: $(git show --format="%s" -q ${k_pending_merge_commit})."
+
+                    # cherry-pick the difference on the filtered mainline
+                    reset-godeps ${f_pending_merge_commit}^1 # unconditionally reset godeps
+                    dst_needs_godeps_update=true
+                    local date=$(git show --format="%aD" -q ${f_pending_merge_commit})
+                    if ! GIT_COMMITTER_DATE="${date}" git cherry-pick --keep-redundant-commits -m 1 ${f_pending_merge_commit} >/dev/null; then
```

Comment:
Do we only cherrypick the merge commit? Where do we cherry pick the feature commit in the master branches?

Reply:
nvm, I saw the merge later.